sendtoroute panic

reproducible panic with accidental hops=[]

```
lncli sendtoroute --payment_hash=4689a97abb7f99a75efed862b3ffffffffffffffffffffff41dee7fa4c937413 --routes='{"routes":[{"total_time_lock":534544,"hops":[],"total_amt_msat":"135000000"}]}'
```


```
panic: runtime error: index out of range

goroutine 61127 [running]:
github.com/lightningnetwork/lnd/vendor/github.com/lightningnetwork/lightning-onion.generateSharedSecrets(0x17713e8, 0x0, 0x0, 0xc427c3f650, 0x11bd13d01c6f432, 0x28258b702057463, 0x3b38f7038c7f83)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightningnetwork/lightning-onion/sphinx.go:224 +0x43b
github.com/lightningnetwork/lnd/vendor/github.com/lightningnetwork/lightning-onion.NewOnionPacket(0x17713e8, 0x0, 0x0, 0xc427c3f650, 0x17713e8, 0x0, 0x0, 0xc4331ccdf8, 0x20, 0x20, ...)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/lightningnetwork/lightning-onion/sphinx.go:285 +0x6a
github.com/lightningnetwork/lnd/routing.generateSphinxPacket(0xc42777bd10, 0xc4331ccdf8, 0x20, 0x20, 0x2, 0x2, 0x40f88e, 0xeb32e0, 0xb4640, 0xd9e640)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:1452 +0x461
github.com/lightningnetwork/lnd/routing.(*ChannelRouter).sendPayment(0xc4202f24e0, 0xc4331ccde0, 0xc42777be00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2, 0xc4311b7500)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:1673 +0x576                                                                                                                                github.com/lightningnetwork/lnd/routing.(*ChannelRouter).SendToRoute(0xc4202f24e0, 0xc42bd91f98, 0x1, 0x1, 0xc4331ccde0, 0x0, 0x0, 0x0, 0x0, 0xc42c0dffc0, ...)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/routing/router.go:1561 +0xa6
main.(*rpcServer).dispatchPaymentIntent(0xc4227a53b0, 0xc4259e7420, 0xc429c080c0, 0x0, 0x0, 0x0, 0x0, 0xc425713258, 0x64e7c0, 0xc42e613c00, ...)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/rpcserver.go:2144 +0x3aa
main.(*rpcServer).sendPayment.func3(0xc429c080c0, 0xc4227a53b0, 0xc4259e7420, 0xc42cbcbea0, 0xc4331ccc00)
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/rpcserver.go:2286 +0xb1
created by main.(*rpcServer).sendPayment
        /home/robban/gocode/src/github.com/lightningnetwork/lnd/rpcserver.go:2277 +0x1d1
```

--- a/routing/router.go
+++ b/routing/router.go
@@ -38,6 +38,13 @@ const (
 	defaultPayAttemptTimeout = time.Duration(time.Second * 60)
 )
 
+var (
+	// ErrNoRouteHopsProvided is returned when a caller attempts to
+	// construct a new sphinx packet, but provides an empty set of hops for
+	// each route.
+	ErrNoRouteHopsProvided = fmt.Errorf("empty route hops provided")
+)
+
 // ChannelGraphSource represents the source of information about the topology
 // of the lightning network. It's responsible for the addition of nodes, edges,
 // applying edge updates, and returning the current block height with which the
@@ -1412,6 +1419,14 @@ func (r *ChannelRouter) FindRoutes(target *btcec.PublicKey,
 // be sent to the first hop within the route.
 func generateSphinxPacket(route *Route, paymentHash []byte) ([]byte,
 	*sphinx.Circuit, error) {
+
+	// As a sanity check, we'll ensure that the set of hops has been
+	// properly filled in, otherwise, we won't actually be able to
+	// construct a route.
+	if len(route.Hops) == 0 {
+		return nil, nil, ErrNoRouteHopsProvided
+	}
+
 	// First obtain all the public keys along the route which are contained
 	// in each hop.
 	nodes := make([]*btcec.PublicKey, len(route.Hops))

--- a/routing/router_test.go
+++ b/routing/router_test.go
@@ -9,14 +9,14 @@ import (
 	"testing"
 	"time"
 
+	"github.com/btcsuite/btcd/wire"
 	"github.com/lightningnetwork/lnd/channeldb"
 	"github.com/lightningnetwork/lnd/htlcswitch"
-	"github.com/btcsuite/btcd/wire"
 
+	"github.com/btcsuite/btcd/btcec"
 	"github.com/davecgh/go-spew/spew"
 	"github.com/lightningnetwork/lightning-onion"
 	"github.com/lightningnetwork/lnd/lnwire"
-	"github.com/btcsuite/btcd/btcec"
 )
 
 // defaultNumRoutes is the default value for the maximum number of routes to
@@ -1939,5 +1939,18 @@ func TestIsStaleEdgePolicy(t *testing.T) {
 	}
 	if ctx.router.IsStaleEdgePolicy(*chanID, updateTimeStamp, 1) {
 		t.Fatalf("router failed to detect fresh edge policy")
+	}
+}
+
+// TestEmptyRoutesGenerateSphinxPacket tests that the generateSphinxPacket
+// function is able to gracefully handle being passed a nil set of hops for the
+// route by the caller.
+func TestEmptyRoutesGenerateSphinxPacket(t *testing.T) {
+	t.Parallel()
+
+	emptyRoute := &Route{}
+	_, _, err := generateSphinxPacket(emptyRoute, testHash[:])
+	if err != ErrNoRouteHopsProvided {
+		t.Fatalf("expected empty hops error: instead got: %v", err)
 	}
 }

